{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Gene Study \n",
    "\n",
    "----\n",
    "\n",
    "\n",
    "Searching online database from NCBI (eg. PubMed, MESH, OMIM, etc.) identify a gene that interests you. We will be using this gene throughout the course, so take time to find something that you would like to analyze in different ways. You do not have to have any previous knowledge about the gene, you will learn everything you need to analyze it as part of class. \n",
    "\n",
    "\n",
    "## Let's Hear About Your Gene\n",
    "In a few sentences, share some basic information about you gene including its function (if there is one), which organism it is found in, and what you found interesting about it. Post it to slack in `#gene`. For example:\n",
    "> \"Bioluminescence refers to the production of light by living organisms. The luxA gene encode the α-subunit of the enzyme luciferase producing the light emitting species. Bioluminescent bacteria are identified in _Vibrionaceae_, _Shewanellaceae_ and _Enterobacteriaceae_ and are mainly found in marine habitats.\"\n",
    "\n",
    "\n",
    "## Let's Hear About Your Gene...Literally\n",
    "We see now, that DNA encodes a tremendous amount of information -- some of which we understand and some that we do not. Many different approaches have been attempted to manipulate and translate these sequences of characters in different ways to try to identify obscured patterns. One of these has been to convert a given DNA sequence into a musical composition.\n",
    "\n",
    "Given a sequence, map each nucleotide base to a musical note and generate a melody based on the DNA sequence provided. Utilize the translated amino acid to generate a complimentary note for the codon. This note should play for all three bases that make up the codon.\n",
    "\n",
    "```\n",
    "♩♩♩ ♩♩♩ \n",
    "TAA CAT\n",
    " ♬   ♬     \n",
    "ILE VAL\n",
    "```\n",
    "\n",
    "There are many approaches to accomplishing this, but the `sounddevice` Python library is a great place to get started. A simple example of using the library is shown in `sounds.py`. Expand on the above code snippet to play the song of your gene and save it as an audio file. Post the audio file along with you gene description.\n",
    "\n",
    "Commit your music composition code and the generated audio file in the repository. Submit the URL for your repository."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sounddevice as sd\n",
    "import numpy as np\n",
    "import soundfile as sf # to save souns file created\n",
    "\n",
    "# Mapping the DNA bases to frequency\n",
    "base_to_frequency = {\n",
    "    'A': 440, 'T': 494, 'G': 523, 'C': 587\n",
    "}\n",
    "\n",
    "# Mapping the amino acid to frequencies (for codons)\n",
    "aa_to_frequency = {\n",
    "    'I': 659, 'V': 698, 'F': 784, 'M': 880, 'L': 988,\n",
    "    'C': 1047, 'A': 1175, 'G': 1319, 'T': 1397,\n",
    "    'S': 1568, 'W': 1760, 'Y': 1976, 'P': 2093,\n",
    "    'H': 2349, 'Q': 2637, 'N': 2794, 'K': 3136,\n",
    "    'D': 3520, 'E': 3951, 'R': 4186\n",
    "}\n",
    "\n",
    "# Mapping codons to amino acids\n",
    "codon_table = {\n",
    "    'ATA': 'I', 'ATC': 'I', 'ATT': 'I', 'ATG': 'M',\n",
    "    'ACA': 'T', 'ACC': 'T', 'ACG': 'T', 'ACT': 'T',\n",
    "    'AAC': 'N', 'AAT': 'N', 'AAA': 'K', 'AAG': 'K',\n",
    "    'AGC': 'S', 'AGT': 'S', 'AGA': 'R', 'AGG': 'R',\n",
    "    'CTA': 'L', 'CTC': 'L', 'CTG': 'L', 'CTT': 'L',\n",
    "    'CCA': 'P', 'CCC': 'P', 'CCG': 'P', 'CCT': 'P',\n",
    "    'CAC': 'H', 'CAT': 'H', 'CAA': 'Q', 'CAG': 'Q',\n",
    "    'CGA': 'R', 'CGC': 'R', 'CGG': 'R', 'CGT': 'R',\n",
    "    'GTA': 'V', 'GTC': 'V', 'GTG': 'V', 'GTT': 'V',\n",
    "    'GCA': 'A', 'GCC': 'A', 'GCG': 'A', 'GCT': 'A',\n",
    "    'GAC': 'D', 'GAT': 'D', 'GAA': 'E', 'GAG': 'E',\n",
    "    'GGA': 'G', 'GGC': 'G', 'GGG': 'G', 'GGT': 'G',\n",
    "    'TCA': 'S', 'TCC': 'S', 'TCG': 'S', 'TCT': 'S',\n",
    "    'TTC': 'F', 'TTT': 'F', 'TTA': 'L', 'TTG': 'L',\n",
    "    'TAC': 'Y', 'TAT': 'Y', 'TAA': '*', 'TAG': '*',\n",
    "    'TGC': 'C', 'TGT': 'C', 'TGA': '*', 'TGG': 'W'\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_tone(frequency, duration_ms=500, sampling_freq=44100):\n",
    "    '''\n",
    "    This function generates a tone for a given frequency (from sounds.py)\n",
    "\n",
    "    Inputs:\n",
    "      frequency, duration_ms, sampling_freq (int): inputs for tone definition\n",
    "\n",
    "    Returns:\n",
    "      tone: the tone for a given frequency\n",
    "    '''\n",
    "    t = np.linspace(0, duration_ms / 1000, int(sampling_freq * duration_ms / 1000), False)\n",
    "    tone = 0.5 * np.sin(2 * np.pi * frequency * t)\n",
    "    return tone\n",
    "\n",
    "def read_fasta(fasta_file):\n",
    "    '''\n",
    "    This function reads gets the DNA sequence from a given the FASTA file.\n",
    "\n",
    "    Inputs:\n",
    "      fasta_file (string): the FASTA file of your gene \n",
    "\n",
    "    Returns:\n",
    "      sequence (string): DNA sequence\n",
    "    '''\n",
    "    with open(fasta_file, 'r') as file:\n",
    "        lines = file.readlines()\n",
    "    \n",
    "    # Ignore the first line if it starts with \">\" (FASTA header)\n",
    "    sequence = ''\n",
    "    for line in lines:\n",
    "        if not line.startswith('>'):\n",
    "            sequence += line.strip()\n",
    "    return sequence\n",
    "    \n",
    "def translate_dna_to_aa(dna_sequence):\n",
    "    '''\n",
    "    This function translates a given DNA sequence into amino acids\n",
    "\n",
    "    Inputs:\n",
    "      dna_sequence (str): the DNA sequence of interest\n",
    "\n",
    "    Returns:\n",
    "      amino_acids (str): the amino acids from the DNA\n",
    "    '''\n",
    "    amino_acids = \"\"\n",
    "    for i in range(0, len(dna_sequence) - 2, 3):\n",
    "        codon = dna_sequence[i:i+3]\n",
    "        amino_acids += codon_table.get(codon.upper(), 'X')  # 'X' for unknown codons\n",
    "    return amino_acids\n",
    "\n",
    "def dna_to_melody(dna_sequence, sampling_freq=44100):\n",
    "    '''\n",
    "    This function converts a DNA sequence to a melody\n",
    "\n",
    "    Inputs:\n",
    "      dna_sequence (str): the DNA sequence of interest\n",
    "      sampling_freq (int): sampling frequency for unknown characters\n",
    "\n",
    "    Returns:\n",
    "      melody (array): updated melody\n",
    "    '''\n",
    "    melody = []\n",
    "    \n",
    "    # Iterate through DNA bases and generate tones\n",
    "    for base in dna_sequence:\n",
    "        freq = base_to_frequency.get(base.upper(), 0)\n",
    "        if freq > 0:\n",
    "            base_tone = generate_tone(freq, duration_ms=250, sampling_freq=sampling_freq)\n",
    "            melody.append(base_tone)\n",
    "        #else: \n",
    "            #print(f\"Base '{base}' not found in the dictionary.\")\n",
    "    \n",
    "    # Translate to amino acids and generate codon tones\n",
    "    amino_acids = translate_dna_to_aa(dna_sequence)\n",
    "    for aa in amino_acids:\n",
    "        freq = aa_to_frequency.get(str(aa), 0)\n",
    "        if freq > 0:\n",
    "            codon_tone = generate_tone(freq, duration_ms=750, sampling_freq=sampling_freq)\n",
    "            melody.append(codon_tone)\n",
    "        #else: \n",
    "            #print(f\"Amino acid '{aa}' not found in the dictionary.\")\n",
    "\n",
    "    return np.concatenate(melody)\n",
    "\n",
    "def save_melody(melody, filename=\"melody.wav\", sampling_freq=44100):\n",
    "    '''\n",
    "    This function saves the melody as a .wav file.\n",
    "    Source: https://python-soundfile.readthedocs.io/en/0.11.0/\n",
    "\n",
    "    Inputs:\n",
    "      melody: \n",
    "      filename (Str): The name you want your sound file to have.\n",
    "      sampling_freq (int): the sampling frequency\n",
    "\n",
    "    Returns:\n",
    "      NA. Saves sound file as.\n",
    "    '''\n",
    "    sf.write(filename, melody, sampling_freq)\n",
    "    print(f\"Melody saved as {filename}\")\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    # Melody for SLC6A4 sequence \n",
    "    fasta_file = \"gene.fna\" # the FASTA file downloaded from NCBI\n",
    "    slc6a4_seq = read_fasta(fasta_file) # get DNA sequence from FASTA file\n",
    "    melody = dna_to_melody(slc6a4_seq) # convert sequence to melody\n",
    "\n",
    "    # save the melody\n",
    "    save_melody(melody, \"SLC6A4_melody_sonia.wav\")\n",
    "    \n",
    "    # Play the melody\n",
    "    #sd.play(melody, samplerate=44100)\n",
    "    #sd.wait()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### References:\n",
    "- Role of Serotonin-Associated Genes: https://www.frontiersin.org/journals/genetics/articles/10.3389/fgene.2020.601868/full\n",
    "- SLC6A4 gene from NCBI: https://www.ncbi.nlm.nih.gov/gene/6532\n",
    "- "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
